src/snowflake/snowpark/modin/plugin/extensions/datetime_index.py

```diff
@@ -0,0 +1,124 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+# Code in this file may constitute partial or total reimplementation, or modification of
+# existing code originally distributed by the Modin project, under the Apache License,
+# Version 2.0.
+
+"""
+Module houses ``DatetimeIndex`` class, that is distributed version of
+``pandas.DatetimeIndex``.
+"""
+
+from __future__ import annotations
+
+import pandas as native_pd
+from pandas._libs import lib
+from pandas._typing import ArrayLike, Dtype, Frequency, Hashable, TimeAmbiguous
+
+from snowflake.snowpark.modin.plugin.compiler.snowflake_query_compiler import (
+    SnowflakeQueryCompiler,
+)
+from snowflake.snowpark.modin.plugin.extensions.index import Index
+
+
+class DatetimeIndex(Index):
+
+    # Equivalent index type in native pandas
+    _NATIVE_INDEX_TYPE = native_pd.DatetimeIndex
+
+    def __init__(
+        self,
+        data: ArrayLike | SnowflakeQueryCompiler | None = None,
+        freq: Frequency | lib.NoDefault = lib.no_default,
+        tz=lib.no_default,
+        normalize: bool | lib.NoDefault = lib.no_default,
+        closed=lib.no_default,
+        ambiguous: TimeAmbiguous = "raise",
+        dayfirst: bool = False,
+        yearfirst: bool = False,
+        dtype: Dtype | None = None,
+        copy: bool = False,
+        name: Hashable | None = None,
+        convert_to_lazy: bool = True,
```

Comment:
@sfc-gh-azhan @sfc-gh-vbudati i am always confused by this convert_to_lazy parameter, is that parameter only used when data is not a SnowflakeQueryCompiler, or even if it is a SnowflakeQueryCompiler, we will convert it to local if convert_to_lazy is true?
And what is the benefit of keeping some index object as local, why not just always pushing it  to remote like what we did for dataframe or series? If the argument is if the data is already in local, then i think the same argument also holds if the given data is not a SnowflakeQueryCompiler

Reply:
It is only used for `df.columns`. Given this can be used quite often, we take this as the only special case to avoid unnecessary round trip.

Reply:
Let's take that offline, there seems some problem for this usage